src/aro/Compilation.zig

```diff
@@ -587,9 +591,18 @@ pub fn generateBuiltinMacros(comp: *Compilation, system_defines_mode: SystemDefi
         try buf.append('\n');
     }
 
-    // timestamps
-    const timestamp = try comp.getTimestamp();
-    try generateDateAndTime(buf.writer(), timestamp);
+    const provided: ?i64 = comp.getSourceEpoch(max_timestamp) catch blk: {
+        try comp.addDiagnostic(.{
+            .tag = .invalid_source_epoch,
+            .loc = .{ .id = .unused, .byte_offset = 0, .line = 0 },
+        }, &.{});
+        break :blk null;
+    };
+    if (provided) |epoch| {
+        try generateDateAndTime(buf.writer(), @intCast(std.math.clamp(epoch, 0, max_timestamp)));
```

Comment:
This clamp could now be moved to `getSourceEpoch` since that's where it's validated.